### PR TITLE
improve evaluation speed for Min and Max with many arguments

### LINKED ISSUE #16249: Min/Max with many args is slow to construct
See https://github.com/sympy/sympy/pull/15617. Something like `Min(*symbols('x:50'))` takes 23 seconds on my computer, and doing `evaluate=False` does not help, because it does not disable `_find_localzeros`. Making `evaluate=False` do less evaluation should be easy to fix. Making the general evaluation faster could be more difficult.

### PR DESCRIPTION
**Pull request by cckolon**
#### References to other Issues or PRs

Fixes #16249 

#### Brief description of what is fixed or changed

Min and Max are currently slow when called with many arguments, mostly because they require pairwise comparisons of all arguments. This PR improves speed by first splitting the arguments into constants, unconstrained symbols, and other values. This allows a fair bit of simplification before the full pairwise comparison, especially when many of the arguments are constants or unconstrained symbols.

I ran the following code on my Macbook before and after the change:
```python
import time

from sympy import symbols, Min

start_time = time.time()
Min(*symbols(names='x:100'))
end_time = time.time()
print(f"Time taken: {end_time - start_time} seconds")
```
before: Time taken: 1.0753211975097656 seconds
after: Time taken: 0.0012416839599609375 seconds

I also added a number of test assertions for Max/Min behavior I was concerned about, and added type annotations to the relevant files to make editing easier in my IDE.

#### Other comments

This is my first PR to sympy! I've been a fan for years. I've tried to follow the guidance here and on the site but please tell me if I've done anything wrong and I'll correct it.

#### AI Generation Disclosure

The code and docstrings are human-written (as is this PR description). I used Claude 5 Opus and Grok 4.5 in Cursor to review my changes and suggest test cases.

#### Release Notes

* functions
  * Sped up Max/Min by pruning the number of pairwise comparisons between arguments

### CODE REVIEW

**Review by JosephMehdiyev — COMMENTED** [review]
I won't (can't) share any opinions on the code changes themselves  as I am not experienced in this part of the code base, but the optimization idea itself looks solid to me

File: sympy/functions/elementary/tests/test_miscellaneous.py [hunk 1]
```diff
@@ -20,6 +20,8 @@
 
 def test_Min():
     from sympy.abc import x, y, z
+    from sympy.functions.elementary.complexes import Abs, re
+    from sympy.functions.elementary.exponential import exp
```
Comment (oscarbenjamin) [inline sympy/functions/elementary/tests/test_miscellaneous.py]: Imports should go at the top of the file.

File: sympy/functions/elementary/miscellaneous.py [hunk 2]
```diff
@@ -568,42 +572,106 @@ def _new_args_filter(cls, arg_sequence):
             else:
                 yield arg
 
-    @classmethod
-    def _find_localzeros(cls, values, **options):
+    @staticmethod
+    def _is_sign_constrained(x: Expr) -> bool:
```
Comment (oscarbenjamin) [inline sympy/functions/elementary/miscellaneous.py]: Why is this changed to a staticmethod?

File: sympy/functions/elementary/miscellaneous.py [hunk 3]
```diff
@@ -568,42 +572,106 @@ def _new_args_filter(cls, arg_sequence):
...
+        check if x is guaranteed to be positive, negative, etc., which
+        implies that we need to consider it when comparing to other,
+        unrelated expressions.
+        """
+        return any(getattr(x, attr) is not None for attr in (
+            'is_positive', 'is_negative',
+            'is_nonnegative', 'is_nonpositive',
+            'is_extended_positive', 'is_extended_negative',
+            'is_extended_nonnegative', 'is_extended_nonpositive',
+        ))
```
Comment (oscarbenjamin) [inline sympy/functions/elementary/miscellaneous.py]: This is checking for `not None` but having any of these be `False` is probably not helpful unless one of them is True.

Also it would be better to spell it out explicitly as `x.is_positive or x.is_negative ...` rather than using `getattr`.

Comment (oscarbenjamin) [reply sympy/functions/elementary/miscellaneous.py]: It is only necessary to check the extended predicates here since those are what determine comparability.

File: sympy/functions/elementary/miscellaneous.py [hunk 4]
```diff
@@ -568,42 +572,106 @@ def _new_args_filter(cls, arg_sequence):
...
+          directly (e.g. `sin(1)**2 + cos(1)**2 > 1` does not resolve).
+        - `unconstrained_symbols`: bare symbols that are not
+          sign-constrained. We don't need to pairwise compare these.
+        - `rest`: all other values
+        """
+        extreme_numbers: set[Expr] = set()
+        unconstrained_symbols: set[Expr] = set()
+        rest: set[Expr] = set()
+        for value in values:
+            if value.is_number:
```
Comment (oscarbenjamin) [inline sympy/functions/elementary/miscellaneous.py]: Why is this checking `is_number`?

File: sympy/functions/elementary/miscellaneous.py [hunk 5]
```diff
@@ -568,42 +572,106 @@ def _new_args_filter(cls, arg_sequence):
...
+                    if comparison == False:
+                        is_new_extreme = False
+                        break
+                    elif comparison == True:
+                        extreme_numbers.remove(extreme_number)
+                    # undecidable: keep both
+                if is_new_extreme:
+                    extreme_numbers.add(value)
+            elif isinstance(value, Symbol) and not cls._is_sign_constrained(value):
+                unconstrained_symbols.add(value)
```
Comment (oscarbenjamin) [inline sympy/functions/elementary/miscellaneous.py:605]: Why is this checking for `Symbol`?

File: sympy/functions/elementary/tests/test_miscellaneous.py (hunk 1 above)
Comment (cckolon) [reply sympy/functions/elementary/tests/test_miscellaneous.py]: Fixed

File: sympy/functions/elementary/miscellaneous.py (hunk 2 above)
Comment (cckolon) [reply sympy/functions/elementary/miscellaneous.py]: This is actually a new method, and it's static because it does not depend on whether the `MinMaxBase` is a `Min` or a `Max`. `_find_localzeros` is still a classmethod

File: sympy/functions/elementary/miscellaneous.py (hunk 3 above)
Comment (cckolon) [reply sympy/functions/elementary/miscellaneous.py]: Fixed here, thanks! I added a test including a tricky edge case too

File: sympy/functions/elementary/miscellaneous.py (hunk 4 above)
Comment (cckolon) [reply sympy/functions/elementary/miscellaneous.py]: All the numbers can be pairwise compared and only the max (or min) kept, which reduces the number of total comparisons. So for example if you have `Min(1, 2, 3, x, y)`, you can first separate it into `(1, 2, 3)` (all numbers) and `(x, y)` (unconstrained symbols), and eliminate all the numbers but `1` before you start comparing with the symbols.

File: sympy/functions/elementary/miscellaneous.py (hunk 5 above)
Comment (cckolon) [reply sympy/functions/elementary/miscellaneous.py:605]: Unconstrained symbols don't need to be compared with each other since one will never dominate the other, so they go into a different set (`unconstrained_symbol_zeros`) in the last loop in `_find_localzeros`, which minimizes the number of times we need to call `_is_connected`.

Comment (oscarbenjamin) [reply sympy/functions/elementary/miscellaneous.py:605]: Why does Symbol need to be treated differently from some other expression that has assumptions?

```python
In [1]: x = symbols('x', positive=True)

In [2]: e = 1 + x**2

In [3]: e.is_positive
Out[3]: True
```

File: sympy/functions/elementary/miscellaneous.py (hunk 4 above)
Comment (oscarbenjamin) [reply sympy/functions/elementary/miscellaneous.py]: Yes, but why `is_number` specifically?

The meaning of `e.is_number -> True` is that `e.evalf()` can evaluate the expression approximately to a real float or complex number.

File: sympy/functions/elementary/miscellaneous.py (hunk 2 above)
Comment (oscarbenjamin) [reply sympy/functions/elementary/miscellaneous.py]: It should be a classmethod. There is never really a reason to use staticmethod.

File: sympy/functions/elementary/miscellaneous.py (hunk 5 above)
Comment (cckolon) [reply sympy/functions/elementary/miscellaneous.py:605]: Imagine this situation

```python
from sympy import symbols, Max

x = symbols('x', negative=True)
y = symbols('y', positive=True)

expr = Max(x, y)  # y
```
Here we can't skip comparing x and y, since there are assumptions about which one is positive. If we separated all symbols, this comparison would never happen and we would get

```python
Max(x, y)
```

So we can only skip comparisons if the variables are unconstrained.

Comment (cckolon) [reply sympy/functions/elementary/miscellaneous.py:605]: Oh, sorry, I think I misunderstood your question. Let me take a look.

Comment (cckolon) [reply sympy/functions/elementary/miscellaneous.py:605]: I think it's still not enough to check all terms for constraints, they need to be both unconstrained _and_ a symbol. Imagine something like this:

```python
from sympy import symbols, Max

x = symbols('x', negative=True)
y = symbols('y', positive=True)

x_expr = x + 1
y_expr = y + 1
print(x_expr.is_negative)  # None, since it could be up to 1

expr = Max(x_expr, y_expr)  # y + 1
```

Here even though `x_expr.is_negative` is `None`, we still need to compare `x_expr` and `y_expr` because `y_expr` is always strictly greater than `x_expr`. In other words, since they're not symbols, there can be other information about their upper and lower bounds besides stuff like `is_positive` or `is_negative`.

File: sympy/functions/elementary/miscellaneous.py (hunk 2 above)
Comment (cckolon) [reply sympy/functions/elementary/miscellaneous.py]: Fixed

File: sympy/functions/elementary/miscellaneous.py (hunk 4 above)
Comment (cckolon) [reply sympy/functions/elementary/miscellaneous.py]: Ah I see, I was using `is_number` as a generic way to say "probably true that x>y can be evaluated to `True` or `False`". Is there a better property than `is_number` to check without actually doing the comparison?

Now that I think about it I could probably reduce LOC by removing `extreme_numbers` and just sorting `rest` by `is_number`, which means all the number comparisons happen first. Will take a look at that shortly.

Comment (cckolon) [reply sympy/functions/elementary/miscellaneous.py]: Just updated, but happy to compare by something better than `is_number` if you think it makes sense.

Comment (cckolon) [reply sympy/functions/elementary/miscellaneous.py]: Just changed `is_number` to `is_comparable` which I think makes more sense

File: sympy/functions/elementary/miscellaneous.py [hunk 6]
```diff
@@ -569,41 +573,90 @@ def _new_args_filter(cls, arg_sequence):
...
-        Sequentially allocate values to localzeros.
+        return bool(
+            x.is_extended_positive or
+            x.is_extended_negative or
+            x.is_extended_nonnegative or
+            x.is_extended_nonpositive
+        )
+
+    @classmethod
+    def _is_more_extreme(cls, x: Expr, y: Expr) -> Boolean:
```
Comment (oscarbenjamin) [inline sympy/functions/elementary/miscellaneous.py]: This method does not seem to be used anywhere.

Comment (oscarbenjamin) [reply sympy/functions/elementary/miscellaneous.py]: Apart from this I think the PR looks good.

Comment (cckolon) [reply sympy/functions/elementary/miscellaneous.py]: Fixed! Yep that was from my earlier implementation which I removed.

**Comment by oscarbenjamin** [comment]
Looks good. Thanks
